acme/acme/jws.py

```diff
@@ -12,14 +12,14 @@
 class Header(jose.Header):
     """ACME-specific JOSE Header. Implements nonce, kid, and url.
     """
-    nonce = jose.Field('nonce', omitempty=True, encoder=jose.encode_b64jose)
-    kid = jose.Field('kid', omitempty=True)
-    url = jose.Field('url', omitempty=True)
+    nonce: bytes = jose.field('nonce', omitempty=True, encoder=jose.encode_b64jose)
+    kid: bytes = jose.field('kid', omitempty=True)
+    url: str = jose.field('url', omitempty=True)
```

Comment:
What makes this of type `str`?

Reply:
Nothing you are right, it is a `bytes`.

Reply:
Actually, were you right the first time?

The only place (these aspects of) this class are used is in this file where `url` and `kid` above have type `Optional[str]` [here](https://github.com/certbot/certbot/blob/267fb94478f2ff68ec0dcb16a66e5e8d88b229ad/acme/acme/jws.py#L51-L52). These fields should be declared with the same type of `Optional[str]` right?

Reply:
In fact I am puzzled. If you look at the implementation `JWS.sign`, you will see that `url` and `kid` are passed to the super-method of `jose.JWS` where the actual type if these two parameters are swallowed by the rest parameter `kwargs: Any` of this method, then go in the limbo of a series of class hierarchy going down to `JSONDeserializable` interface. I have basically no idea of what should be the acceptable types going on that direction (callee hierarchy).

So if I go on the opposite direction, we can see from the unit tests for `acme.jws` that `url` and `kid` are instantiated as string. I will trust that then.

For the optional part, from what I understood `omitempty=true` will make the field take the value of the parameter `default` if not set. Given that we do not pass a default parameter, then it can be `None` which make sense with the signature of `sign`.

By the way `nonce` is `Optional[bytes]` following this logic.

Reply:
Arg. Incompatible types with the superclass `jose.Header` ...

Reply:
Fixed here with a TODO if merged before https://github.com/certbot/josepy/pull/122.